recursive_diff/recursive_diff.py

```diff
@@ -318,15 +318,27 @@ def diff(msg, print_path=path):
                 diffs = (lhs.astype('<M8[ns]').astype(int)
                          != rhs.astype('<M8[ns]').astype(int))
 
+            elif lhs.dtype.kind != rhs.dtype.kind:
+                # The cases of comparison between different numeric types
+                # should have been captured in the blocks above.  As far as
+                # numpy's concern, values between diffeent non-numeric types
+                # are not compatible and therefore not equal.
+                diffs = numpy.full(lhs.shape, dtype=bool, fill_value=True)
+
             else:
-                # At least one between lhs and rhs is non-numeric,
-                # e.g. bool or str
-                diffs = lhs.values != rhs.values
-
-                # Comparison between two non-scalar, incomparable types
-                # (like strings and numbers) will return True
-                if diffs is True:
-                    diffs = numpy.full(lhs.shape, dtype=bool, fill_value=True)
+                # Now we are dealing with same (but non-numeric and
+                # non-datetime) dtype comparison.  Ideally we should use
+                # numpy.not_equal() function; however, currently (early 2019)
+                # it returns a NotImplemented object for non-numeric dtypes
+                # (and therefore does not performing the required element-wise
+                # comparsion).  The numpy.not_equal() function in such case
+                # would also promise it will be fix this problem in the future
+                # by giving a FutureWarning, which suppression requires further
+                # handling.  Let's perform the element-wise comparison directly
+                # in python to avoid these numpy function deficiencies.
+                diffs = numpy.array(
+                    [lve != rve for lve, rve in zip(lhs.values, rhs.values)],
+                    dtype=bool).reshape(lhs.shape)
```

Comment:
Are there unit tests for bool and str? Are they currently raising the FutureWarning? What other use cases are we dealing with?

Reply:
Okay, i added a new file test_recursive_diff_numpy_types.py.  I put all the test cases that I think likely to impact together in this file so that I can switch between branches and run the same tests.  All test cases passed with original code (with some warnings), and also passed with my latest proposal (with no warnings).  If you are satisfied with the test cases we can either leave this new test file alone, or merge the new test file back to test_recursive_diff.py.  Let me know if you spot that I miss some other test cases.